07-Threads/threads.c

```diff
@@ -49,23 +49,37 @@ void systick_handler()
 	*SCB_ICSR |= SCB_ICSR_PENDSVSET;
 }
 
-void thread_start()
+void __attribute__((naked)) thread_start()
 {
 	lastTask = 0;
 
-	/* Save kernel context */
-	asm volatile("mrs ip, psr\n"
-	             "push {r4-r11, ip, lr}\n");
-
+	/* Reset APSR before context switch */
+	asm volatile("mov r0, #0\n"
+	             "msr APSR_nzcvq, r0\n");
 	/* To bridge the variable in C and the register in ASM,
 	 * move the task's stack pointer address into r0.
 	 * http://www.ethernut.de/en/documents/arm-inline-asm.html
 	 */
 	asm volatile("mov r0, %0\n" : : "r" (tasks[lastTask].stack));
-	/* Load user task's context and jump to the task */
-	asm volatile("ldmia r0!, {r4-r11, lr}\n"
-                    "msr psp, r0\n"
-                    "bx lr\n");
+	asm volatile("msr psp, r0\n"
+	             "mov r0, #3\n"
+	             "msr control, r0\n"
+	             "isb\n");
+	/* This is how we simulate stack handling that pendsv_handler
+	 * does. Thread_create sets 17 entries in stack, and the 9
```

Comment:
Can you illustrate with some text-based diagram(s)?

Reply:
Yes, as this diagram is accepted, I can copy to 08-CMSIS or move to a document (although there is no such file for document now)

Reply:
If the diagram is okay, I'll copy to 08-CMSIS?

Reply:
Please go ahead.